### PR TITLE
release-21.2: release-22.1: backupccl: elide spans from backups that were subsequently reintroduced

### PR DESCRIPTION
**Pull request by msbutler**
Backport 2/2 commits from #88488.

/cc @cockroachdb/release

---

Backport 1/1 commits from #87312.

/cc @cockroachdb/release

---

Currently RESTORE may restore invalid backup data from a backed up table that
underwent an IMPORT rollback. See https://github.com/cockroachdb/cockroach/issues/87305 for a detailed explanation.

This patch ensures that RESTORE elides older backup data that were deleted via
a non-MVCC operation. Because incremental backups always reintroduce spans
(i.e. backs them up from timestamp 0) that may have undergone a non-mvcc
operation, restore can identify restoring spans with potentially corrupt data
in the backup chain and only ingest the spans' reintroduced data to any system
time, without the corrupt data.

Here's the basic impliemenation in Restore:
- For each span we want to restore
   - identify the last time, l, the span was introduced, using the manifests
   - dont restore the span using a backup if backup.EndTime < l

This implementation rests on the following assumption: the input spans for each
restoration flow (created in createImportingDescriptors) and the
restoreSpanEntries (created by makeSimpleImportSpans) do not span across
multiple tables. Given this assumption, makeSimpleImportSpans skips adding
files from a backups for a given input span that was reintroduced in a
subsequent backup.

It's worth noting that all significant refactoring occurs on code run by
the restore coordinator; therefore, no special care needs to be taken for
mixed / cross version backups. In other words, if the coordinator has updated,
the cluster restores properly; else, the bug will exist on the restored cluster.
It's also worth noting that other forms of this bug are apparent on older
cluster versions (https://github.com/cockroachdb/cockroach/issues/88042, https://github.com/cockroachdb/cockroach/issues/88043) and has not been noticed by customers; thus,
there is no need to fail a mixed version restore to protect the customer from
this already existing bug.

Informs https://github.com/cockroachdb/cockroach/issues/87305

Release justification: bug fix

Release note: fix for TA advisory
https://cockroachlabs.atlassian.net/browse/TSE-198

### CODE REVIEW

**Comment by msbutler** [comment]
The significant changes relative to the 22.1 patch are the use of aditya's fixtures to generate backups for the unit tests. So, the refactoring related to test infra and the two dd tests deserve a closer look.

I also plan to squash commits 3 and 4 into 2 and delete commit 5.

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-missing-data [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+----
+1
+
+query-sql
+SELECT count(*) FROM d.goodfoo;
+----
+3
+
+
+# Check that that backup chain with the intermediate incremental backup restores properly
```
Comment (adityamaru) [inline pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-missing-data:188]: nit: I think we need to say backup without revision history here?

**Review by adityamaru — DISMISSED** [review]
Mostly LGTM, but I've only given the datadriven tests a close review. I'm assuming the rest backported relatively cleanly.

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-missing-data [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+HINT: See: https://www.cockroachlabs.com/docs/advisories/a88042
+
+
+# Ensure the restore succeeds if there are no corrupt tables in the target
+exec-sql
+RESTORE TABLE d.baz FROM LATEST IN 'nodelocal://0/database/' with into_db = defaultdb;
+----
+
+
+# Check that the wildcard restore encounters the same errors
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-missing-data]: Nah. I think we're sufficiently covered.

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-missing-data (hunk 1 above)
Comment (msbutler) [reply pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-missing-data]: ah. good catch!

File: pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-rollback [hunk 3]
```diff
@@ -0,0 +1,137 @@
+
+# Test that backup fixtures created using the following timeline are properly restored
+# - begin import jobs and pause them
+# - run inc backup - verify inc has captured the introduced spans (unlike the fixtures in
+#   in-progress-missing-data)
+# - roll it back it back with non-mvcc clear range
+# - run an inc backup and ensure we reintroduce the table spans
+# (the fixtures are located in ccl/backupccl/testdata/backup-restore/restore_importing_tables_with_patch)
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/testdata/backup-restore/in-progress-import-rollback:8]: Done. dirs now roughly correspond to dd test file name.

**Comment by msbutler** [comment]
I've added a commit that filters views out of restore's checker, though perhaps it's better for this commit to get incorperated in https://github.com/cockroachdb/cockroach/pull/86713

**Comment by msbutler** [comment]
i plan to squash commits 3 and 4 into commit 2.

**Review by stevendanna — APPROVED** [review]

**Pull request merged**